### PR TITLE
Add neutral_venue option to exclude home advantage per match

### PR DESCRIPTION
**Pull request by ssoulless**
Adds an optional per-match `neutral_venue` flag to the six frequentist goal models. When a match is flagged neutral, the home advantage term is excluded from its expected goals during fitting, so `home_advantage` is estimated only from genuine home games.

This matters for tournaments played at neutral venues (World Cups, continental cups): labelling one side "home" otherwise dilutes the home advantage estimate.

## Changes
  - Optional `neutral_venue` argument on all six frequentist models (Poisson, Dixon-Coles, Negative Binomial, ZIP, Bivariate Poisson, Weibull Copula).
  - Threaded through the Cython loss and gradient kernels: `home_advantage` enters `λ_home` only when `neutral_venue == 0`.
  - Length and 0/1 value validation in `BaseGoalsModel`.

## Backwards compatibility
`neutral_venue` defaults to `None`. Omitting it, passing `None`, or passing an all-zeros array is bit-identical to previous behaviour — existing callers are unaffected.

## Tests
New `test/test_neutral_venue.py` covering all six models: backwards compatibility, loss/gradient response to the flag, the home-advantage gradient being zero for neutral matches, all-neutral convergence, and input validation.

### CODE REVIEW

**Comment by ssoulless** [comment]
@martineastwood Just added to bayesian models as well. This is specially for Worldcup. I've been using it for a while now and I think it's ready to merge it to the library, let me know if there's anything you want to discuss about.

**Comment by martineastwood** [comment]
Thanks for the PR. I like the direction and would be happy to include it once it is complete, but there are a couple of design issues to address first.

Taking poisson_loss_function as the simplest example: you now pass neutral_venue through to the Cython loss, and if neutral_venue[i] == 1 for every fixture then the home rate correctly reduces to

lambda_home = exp(attack[home_idx] + defence[away_idx])

because hfa * (1 - 1) = 0.

The consequence of that though is during fitting:

- hfa drops out of the likelihood entirely
- the objective is flat with respect to hfa
- the gradient with respect to hfa is zero
- the data provides no information about hfa

So the optimiser is still carrying a hfa parameter, but it is not identified. In an all-neutral dataset, any value of hfa is then equally valid from the fit’s point of view.

The problem then appears at prediction time: the current prediction path still applies hfa, so the model can end up using an essentially arbitrary home-advantage value in its forecasts.

I think the pragmatic approach is:

- add neutral_venue to the prediction API
- thread it through prediction in the same way it is threaded through the loss functions

For example:

- predict(home_team, away_team, max_goals=15, normalize=True, neutral_venue=False)
- predict_many(..., neutral_venue=None) where neutral_venue is array-like 0/1

Then, if all training matches are neutral, set hfa = 0 after fitting.

Please can you also update the relevant docs/examples as part of the PR so the new neutral_venue behavior is documented end-to-end.

I think that should then give this feature the correct implementation.

**Comment by ssoulless** [comment]
Thanks for the thorough review, the identifiability point is well taken.

Addressed in the latest commits:
  - `predict` and `predict_many` now accept `neutral_venue` (scalar bool / array) and exclude home advantage from the prediction, mirroring the loss path.
  - When every training match is neutral, `home_advantage` is pinned to 0 after fitting, so the unidentified parameter can't leak into forecasts.
  - Both apply to all goal models, including the Bayesian ones.
  - Documented end-to-end in `docs/models/overview.rst` (training + prediction).

**Comment by martineastwood** [comment]
Thanks, this is looking good for the MLE models.

There is still an issue for the Bayesian models, though. _zero_home_advantage_if_all_neutral() sets the fitted home_advantage parameter to zero, but at inference time the Bayesian models integrate over self.trace, and that trace is not zeroed. As a result, if the training data is all neutral, the model can still apply a non-zero home advantage at prediction time even though the fitted parameter reports 0.0.

The fix would be to zero the home-advantage column in self.trace as well when all training matches are neutral, or otherwise make Bayesian prediction ignore it in that case.

A useful test here would be: after fitting a Bayesian model on all-neutral data, predict(..., neutral_venue=False) should match predict(..., neutral_venue=True).

**Comment by ssoulless** [comment]
fixed in the latest commits + a bug found in the hierarchical bayesian model:

  - `_zero_home_advantage_if_all_neutral` now also zeros the home advantage column of `self.trace` for the Bayesian models, so prediction integrates over a trace whose hfa is actually 0 (not just `self._params`).
  - While addressing this I also noticed a pre-existing indexing bug in `bayesian_predict_c`: it was reading `hfa = trace[s, n_params-2]` and `rho = trace[s, n_params-1]`. That's correct for `BayesianGoalModel` but for `HierarchicalBayesianGoalModel` those positions hold `sigma_attack` and `sigma_defense`. The hierarchical model's predictions were silently reading the sigmas as hfa/rho. Changed to `2 * n_teams` and `2 * n_teams + 1` to match how `_bayesian_log_prob_c` / `_hierarchical_log_prob_c` already index, heads up that hierarchical prediction numbers will change.
  - Your suggested test (`predict(neutral_venue=False) == predict(neutral_venue=True)` after all-neutral fit) is in `test/test_neutral_venue.py`, parametrized over both Bayesian models. Passes for both with the fixes above.

**Comment by martineastwood** [comment]
Thank you, I’ll give it a final review and then get it merged in and released next weekOn 23 May 2026, at 02:36, Sebastian Velandia ***@***.***> wrote:﻿ssoulless left a comment (martineastwood/penaltyblog#23)
fixed in the latest commits + a bug found in the hierarchical bayesian model:

_zero_home_advantage_if_all_neutral now also zeros the home advantage column of self.trace for the Bayesian models, so prediction integrates over a trace whose hfa is actually 0 (not just self._params).
While addressing this I also noticed a pre-existing indexing bug in bayesian_predict_c: it was reading hfa = trace[s, n_params-2] and rho = trace[s, n_params-1]. That's correct for BayesianGoalModel but for HierarchicalBayesianGoalModel those positions hold sigma_attack and sigma_defense. The hierarchical model's predictions were silently reading the sigmas as hfa/rho. Changed to 2 * n_teams and 2 * n_teams + 1 to match how _bayesian_log_prob_c / _hierarchical_log_prob_c already index, heads up that hierarchical prediction numbers will change.
Your suggested test (predict(neutral_venue=False) == predict(neutral_venue=True) after all-neutral fit) is in test/test_neutral_venue.py, parametrized over both Bayesian models. Passes for both with the fixes above.
